### PR TITLE
[MLX] Add off-graph KV cache export mode for HF models

### PR DESCRIPTION
**Pull request by kiymetakdemir**
**Summary**

Adds --use-offgraph-cache, which exports a HuggingFace causal LM against kvcache::update_and_attend instead of an in-graph cache. The model runs with use_cache=False and past_key_values=None, so each attention layer emits one op fed only that step's k/v; history lives in a cache the runtime owns and binds by cache_key. KV-sharing layers address their donor's cache rather than one of their own, so gemma-4 E2B needs 15 caches for its 35 layers.

  Files

  - hf_attention.py — registers the mlx_offgraph attention implementation and a mask function returning None, since the op masks internally from the position it is given. OffGraphExportWrapper exposes the (input_ids, cache_position) signature a runner drives and passes position_ids explicitly, because HuggingFace otherwise derives them from past_key_values.get_seq_length(), which is
  - export_llm_hf.py — adds the off-graph export path and publishes the layout as get_n_caches, get_kv_heads, get_head_dims, get_windows.
  - cache.py — reshapes the HFStaticCache fallback's cache_position to 1-D. That path belongs to the in-graph cache, but it is the fix gemma-4 needs: gemma-4 calls update() without cache_kwargs, so the fallback reads a 0-dim cumulative_length and torch.export fails indexing it.

  Test

  Exported Llama-3.2-1B and gemma-3-1b; both partition into a single MLX subgraph, and the published layout matches each architecture — 16 flat caches for llama, 26 for gemma-3 with full attention at layers 5/11/17/23.

  python -m executorch.backends.mlx.examples.llm.export_llm_hf \
      --model-id unsloth/gemma-3-1b-it --output gemma3_offgraph.pte \
      --use-offgraph-cache --max-seq-len 2048 --dtype fp32

  python -m executorch.backends.mlx.examples.llm.export_llm_hf \
      --model-id unsloth/Llama-3.2-1B-Instruct --output llama_offgraph.pte \
      --use-offgraph-cache --max-seq-len 2048 --dtype bf16

  Add --qlinear 4w for INT4.

### CODE REVIEW

**Comment by metascroy** [comment]
@claude review this code

**Comment by kiymetakdemir** [comment]
@claude review this code

**Comment by metascroy** [comment]
@claude review this code

File: backends/mlx/llm/hf_attention.py [hunk 1]
```diff
@@ -87,6 +102,74 @@ def mlx_sdpa_with_start_pos_forward(
...
+    module: torch.nn.Module,
+    query: torch.Tensor,  # [B, num_heads, q_len, head_dim] - BHSD
+    key: torch.Tensor,  # [B, num_kv_heads, q_len, head_dim] - BHSD (this step)
+    value: torch.Tensor,  # [B, num_kv_heads, q_len, head_dim] - BHSD (this step)
+    attention_mask: Union[torch.Tensor, "BlockMask"],  # noqa: F821
+    position_ids: Optional[torch.Tensor] = None,
+    scaling: Optional[float] = None,
+    softcap: Optional[float] = None,
+    head_mask: Optional[torch.Tensor] = None,
+    **kwargs,
```
Comment (metascroy) [inline backends/mlx/llm/hf_attention.py:143]: Is it better to have **kwargs catch everything else (including softcap), and then raise if kwargs is non-empty? That way more unknown kwargs don't sneak in

Comment (kiymetakdemir) [reply backends/mlx/llm/hf_attention.py:143]: It makes sense for catching unknown ones, I added a sweep that pops the args we handle elsewhere (is_causal, use_cache, sliding_window, dropout) and raises on anything left. But softcap and head_mask still need to be named, models pass them often as None which still would raise.

**Comment by kiymetakdemir** [comment]
For claude review;
1. Dim's default min is 0, not 2 (torch/export/dynamic_shapes.py:180), so the constraint is in range [0, max_seq_len-1] and q_len == 1 is verified in practice, since the runner decodes one token per step. 
2. Added assert input_ids.shape[0] == 1 in the export wrapper.
3. I'll add MaskKind.EXPLICIT in a separate PR.
4. The value is uniform because transformers exposes sliding_window as a scalar on the text config; layer_types only says which layers are sliding, so there's no per-layer window to read today.
5. Added a comment at the assignment stating that get_n_caches is the cache count rather than num_hidden_layers.
6. Extracting it would mean touching the other export paths, which are out of scope for this PR.

**Review by metascroy — APPROVED** [review]
